### PR TITLE
Log errors on `uncaughtException` `unhandledRejection` events

### PR DESCRIPTION
**Pull request by ismail-syed**
## Description
See title

Fixes https://github.com/Shopify/quilt/issues/958

- [x]  Minor: New feature (non-breaking change which adds functionality)

I did a 🎩 in [`web-rails-proving-ground`](https://github.com/Shopify/web-rails-proving-ground/pull/164) with this change along with a change to generate a [exception](https://github.com/Shopify/quilt/pull/react-server-uncaught-server-exceptions-test-release).

Shipit deploy [logs](https://shipit.shopify.io/shopify/web-rails-proving-ground/production/deploys/849980) contain more meaningful logs on why the node server failed on startup.

```
[FATAL][2019-09-26 14:07:54 +0000]	      $ node build/server/main.js
[FATAL][2019-09-26 14:07:54 +0000]	      React Server failed to start.
[FATAL][2019-09-26 14:07:54 +0000]	      Error: Test error. I am an uncaught exception thrown in the server that shows in shipit logs
[FATAL][2019-09-26 14:07:54 +0000]	      at breakThings (/app/build/server/webpack:/app/ui/server.js:18:15)
[FATAL][2019-09-26 14:07:54 +0000]	      at Module.breakThings (/app/build/server/webpack:/app/ui/server.js:34:5)
```  

## Checklist

- [x] I have added a changelog entry, prefixed by the type of change noted above

### CODE REVIEW

File: packages/react-server-webpack-plugin/src/react-server-webpack-plugin.ts [hunk 1]
```diff
@@ -81,6 +81,13 @@ function serverSource(options: Options) {
     import React from 'react';
     import {createServer} from '@shopify/react-server';
     import App from 'index';
+    process.on('uncaughtException', logError);
+    process.on('unhandledRejection', logError);
+    function logError(error) {
+      const errorLog = \`\${error.stack || error.message || 'No stack trace was present'}\`;
+      console.log(\`React Server failed to start.\n\${errorLog}\`);
```
Comment (GoodForOneFare) [inline packages/react-server-webpack-plugin/src/react-server-webpack-plugin.ts]: How confident are you about this message only being thrown during startup?  Does koa catch 100% of rejected promises?

Comment (ismail-syed) [reply packages/react-server-webpack-plugin/src/react-server-webpack-plugin.ts]: I'm pretty confident about this change. I tried creating some e2e tests which would involve testing dev startup which gets too messy with our e2e testing approach. Unless there's an intuitive way for us to test dev startups that I'm not aware of. 


> Does koa catch 100% of rejected promises

No, not natively to koa. However, our [logger](https://github.com/Shopify/quilt/blob/master/packages/react-server/src/logger/logger.ts#L88-L91) will handles rejected promises within koa.

**Comment by GoodForOneFare** [comment]
I'm a little confused that this change didn't happen in [`templates/server.ts`](https://github.com/Shopify/quilt/blob/master/packages/react-server-webpack-plugin/src/templates/server.ts).  Are those templates not used, yet?

**Comment by ismail-syed** [comment]
> Are those templates not used, yet?

This drove me crazy for longer that I'd like 😅 ... Turns out [they're not](https://github.com/Shopify/quilt/pull/1021).

**Pull request merged**